### PR TITLE
[4.0][com_content] add duplicate from article actions dropdown 

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #28864 .

### Summary of Changes
added a duplicate item from article actions dropdown 

### Testing Instructions
content->articles
select an article
click on actions dropdown 
click on duplicate

Screenshot from 2020-04-30 12-36-28

### Expected result
you can duplicate an article with few clicks

### Actual result
N/A

### Documentation Changes Required
maybe

### CODE REVIEW

**Comment by brianteeman** [comment]
If you cannot duplicate multiple articles then the option should be disabled when multiple articles are selected. Otherwise we get this message with untranslated string

image

**Comment by brianteeman** [comment]
Apart from the missing strings shouldnt this also be a success not an info ?
image

**Comment by SharkyKZ** [comment]
No reason to limit this to 1 article.

**Comment by SharkyKZ** [comment]
ACL checks are missing.

**Comment by alikon** [comment]
if you want to act on multiple articles you need to use batch imho

**Comment by alikon** [comment]
still haven't found a way to restrict action on a single item 
any help ?

**Comment by SharkyKZ** [comment]
Why would you limit this to 1 article? Modules don't have such restriction.

**Comment by alikon** [comment]
> ACL checks are missing.

are the same as for bacth

https://github.com/joomla/joomla-cms/pull/28881/files#diff-03b2b33621298e1452643642d4661f3bR237

**Comment by alikon** [comment]
> Why would you limit this to 1 article? Modules don't have such restriction.

make little sense to duplicate more than 1 to me or  better there is the batch mode for that
or i've fully misinterpreted the issue  #28864 .... reduce the number of clicks

**Comment by alikon** [comment]
> Apart from the missing strings shouldnt this also be a success not an info ?

sure

**Comment by SharkyKZ** [comment]
Batch requires changing category for copy to work. So it's not the same as duplicating.

>are the same as for bacth

Those are just cosmetic. The checks in the controller are missing.

**Comment by softforge** [comment]
Agree this would be a very useful feature. Batch is used to change a group of articles from one thing to another. This is copying an article or group of articles using the category, user group etc of the original. Ideal if you have set up template articles for clients and you want them to leave the template ones, copy them and then work on them

**Comment by brianteeman** [comment]
That's exactly the use case for Save as Copy. Except you're adding an extra step

File: administrator/components/com_content/src/View/Articles/HtmlView.php [hunk 1]
```diff
@@ -233,7 +233,7 @@ protected function addToolbar()
 				$childBar->trash('articles.trash')->listCheck(true);
 			}
 
-			// Add a batch button
+			// Add a batch and duplicate button
```
Comment (Quy) [inline administrator/components/com_content/src/View/Articles/HtmlView.php]:
```suggestion
			// Add batch and duplicate buttons
```

**Comment by sere-b** [comment]
I have tested this item :white_check_mark: successfully on afc022b55e602a744905a862330f9849cf86e386<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28881">issues.joomla.org/tracker/joomla-cms/28881</a>.</sub>

**Comment by SharkyKZ** [comment]
Tags and featured data is missing.

File: administrator/components/com_content/src/Model/ArticleModel.php [hunk 2]
```diff
@@ -1517,4 +1517,40 @@ public function runTransition(int $pk, int $transition_id): bool
...
+		$article['asset_id'] = null;
+		list($title, $alias) = $this->generateNewTitle($article['catid'], $article['alias'], $article['title']);
+		$article['title']    = $title;
+		$article['alias']    = $alias;
+		$tags                = new TagsHelper;
+		$article['tags']     = (array) $tags->getTagIds($pk, 'com_content.article');
+
+		if (!$this->save($article))
+		{
+			throw new \Exception(Text::_('JLIB_DATABASE_ERROR_ARTICLE_UNIQUE_ALIAS'));
```
Comment (SharkyKZ) [inline administrator/components/com_content/src/Model/ArticleModel.php:1553]: Don't assume this.

File: administrator/components/com_content/src/Controller/ArticlesController.php [hunk 3]
```diff
@@ -159,4 +159,54 @@ public function getQuickiconContent()
...
+			{
+				throw new \Exception(Text::_('COM_CONTENT_NO_ITEM_SELECTED'));
+			}
+
+			if (\count($pks) > 1)
+			{
+				throw new \Exception(Text::_('COM_CONTENT_MORE_ITEMS_SELECTED'));
+			}
+
+			if (! (($user->authorise('core.create', 'com_content.article.' . (int) $pks)
```
Comment (SharkyKZ) [inline administrator/components/com_content/src/Controller/ArticlesController.php:191]: Don't cast to int. It always results in 1 here.

File: administrator/components/com_content/src/Model/ArticleModel.php [hunk 4]
```diff
@@ -1517,4 +1517,40 @@ public function runTransition(int $pk, int $transition_id): bool
...
+
+		$article = $db->setQuery($query)->loadAssoc();
+
+		$article['id']       = null;
+		$article['asset_id'] = null;
+		list($title, $alias) = $this->generateNewTitle($article['catid'], $article['alias'], $article['title']);
+		$article['title']    = $title;
+		$article['alias']    = $alias;
+		$tags                = new TagsHelper;
+		$article['tags']     = (array) $tags->getTagIds($pk, 'com_content.article');
```
Comment (SharkyKZ) [inline administrator/components/com_content/src/Model/ArticleModel.php:1549]: You are casting a comma separated string to array.

File: administrator/components/com_content/src/Model/ArticleModel.php [hunk 5]
```diff
@@ -1517,4 +1517,40 @@ public function runTransition(int $pk, int $transition_id): bool
...
+	 * Method to duplicate an article.
+	 *
+	 * @param   array  $pk  An array of primary key IDs.
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 * @throws  \Exception
+	 */
+	public function duplicate(Array $pk): void
```
Comment (SharkyKZ) [inline administrator/components/com_content/src/Model/ArticleModel.php:1531]: It you're really going to limit this to 1 item, make `$pk` integer. But, again, this makes no sense.

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on c2d6077edf2a451ae0d9a7677b42d085180feb31<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28881">issues.joomla.org/tracker/joomla-cms/28881</a>.</sub>

**Comment by adj9** [comment]
I have not tested this item.<br /><br />It's a very good idea for similar articles per do last articles.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28881">issues.joomla.org/tracker/joomla-cms/28881</a>.</sub>

**Comment by jwaisner** [comment]
@alikon can you please fix conflicts?

**Pull request closed without merging**